gcm/benchmarks/benchmarking_test.go

```diff
@@ -0,0 +1,300 @@
+}
+
+func BenchmarkGCM_8Workers100MilliTimeout(b *testing.B) {
+	params := &benchParams{
+		B:             b,
+		workers:       8,
+		subscriptions: 8,
+		timeout:       100 * time.Millisecond,
+		clients:       8,
+		sender:        sendMessageSample,
+	}
+	params.throughputSend()
+	fmt.Println(params)
+}
+
+func BenchmarkGCM_16Workers100MilliTimeout(b *testing.B) {
+	params := &benchParams{
+		B:             b,
+		workers:       16,
+		subscriptions: 8,
+		timeout:       100 * time.Millisecond,
+		clients:       8,
+		sender:        sendMessageSample,
+	}
+	params.throughputSend()
+	fmt.Println(params)
+}
+
+func sendMessageSample(c client.Client) error {
+	return c.Send(gcmTopic, "test-body", "{id:id}")
+}
+
+type sender func(c client.Client) error
+
+type benchParams struct {
+	*testing.B
+	workers       int           // number of gcm workers
+	subscriptions int           // number of subscriptions listening on the topic
+	timeout       time.Duration // gcm timeout response
+	clients       int           // number of clients
+	sender        sender        // the function that will send the messages
+	sent          int           // sent messages
+	received      int           // received messages
+
+	service  *server.Service
+	receiveC chan bool
+	doneC    chan struct{}
+
+	wg    sync.WaitGroup
+	start time.Time
+	end   time.Time
+}
+
+func (params *benchParams) String() string {
+	return fmt.Sprintf(`
+		Throughput %.2f messages/second using:
+			%d workers
+			%d gcm subscriptions
+			%s gcm response timeout
+			%d clients
+	`, params.mps(), params.workers, params.subscriptions, params.timeout, params.clients)
+}
+
+func (params *benchParams) expectedMessagesCount() int {
+	return params.N * params.clients * params.subscriptions
+}
+
+func (params *benchParams) send(c client.Client) error {
+	err := params.sender(c)
+	if err != nil {
+		return err
+	}
+	params.sent++
+	return nil
+}
+
+func (params *benchParams) receiveLoop() {
+	for i := 0; i <= params.workers; i++ {
+		go func() {
+			for {
+				select {
+				case <-params.receiveC:
+					params.received++
+					logger.WithField("received", params.received).Info("Received gcm call")
+					params.wg.Done()
+				case <-params.doneC:
+					return
+				}
+			}
+		}()
+	}
+}
+
+// start the service
+func (params *benchParams) setUp() {
+	params.doneC = make(chan struct{})
+	params.receiveC = make(chan bool)
+
+	a := assert.New(params)
+
+	dir, errTempDir := ioutil.TempDir("", "guble_benchmarking_gcm_test")
+	defer func() {
+		errRemove := os.RemoveAll(dir)
+		if errRemove != nil {
+			logger.WithFields(log.Fields{"module": "testing", "err": errRemove}).Error("Could not remove directory")
+		}
+	}()
+	a.NoError(errTempDir)
+
+	*config.Listen = "localhost:0"
+	*config.KVBackend = "memory"
+	*config.MSBackend = "file"
+	*config.StoragePath = dir
+	*config.GCM.Enabled = true
+	*config.GCM.APIKey = "WILL BE OVERWRITTEN"
+	*config.GCM.Workers = params.workers
+
+	params.service = gubled.StartService()
```

Comment:
Since you start the gubled service it will be a good idea to call stop() on this service in order to be sure that the ports used by gubled will be free